### PR TITLE
Details in json format

### PR DESCRIPTION
**Pull request by agunnarsson**
This adds json fromat for details request. Some of the details has been made optional since loading is a bit expensive. 
New parameters for details are &addressdetails=1&linkedplaces=1&childplaces=1 default are false for all parameters details, but they have been included in links on html pages.

### CODE REVIEW

**Comment by mtmail** [comment]
I like it.

One long-term goal is to split off the HTML frontend and let it talk to Nominatim only via the API (JSON/XML output). The reasoning is to make this project easier to manage in terms of writing tests, releasing or having the frontend in a different programming language. Your pull request is one step in that direction.

**Comment by agunnarsson** [comment]
Sounds good! I did not include all information from the html version for now, but it's an easy task if/when the HTML frontend is updated.

File: lib/template/details-html.php [hunk 1]
```diff
@@ -161,6 +161,8 @@ function _one_keyword_row($keyword_token,$word_id){
 					if ($aPointDetails['wikipedia'])
 					{
 						kv('Wikipedia Calculated' , wikipedia_link($aPointDetails) );
+					} else if ($aPointDetails['aExtraTags']['wikipedia']) {
+						kv('Wikipedia Calculated' , wikipedia_link($aPointDetails['aExtraTags']) );
```
Comment (lonvia) [inline lib/template/details-html.php]: That's wrong as the wikipedia tag in ExtraTags is different. That is the originally tagged value. 'Wikipedia Calculated' is the wikipedia page that has been used to compute the importance.

File: lib/template/details-json.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+	function wikipedia_link($sWikipedia)
+	{
+		if ($sWikipedia)
+		{
+			list($sWikipediaLanguage,$sWikipediaArticle) = explode(':',$sWikipedia);
+			return 'https://'.$sWikipediaLanguage.'.wikipedia.org/wiki/'.urlencode($sWikipediaArticle);
+		}
+		return '';
+	}
```
Comment (lonvia) [inline lib/template/details-json.php]: code duplication from details-html.php. It's time to start a new file lib/output.php and move all the functions from details-*.php in there.

File: lib/template/details-json.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	$aPlace['extra_tags'] = $aPointDetails['aExtraTags'];
+
+	if ($aAddressLines) {
+		$aAddressDetails = array();
+		foreach($aAddressLines as $aAddressLine)
+		{
+			$aAddressDetails[] = getDetails($aAddressLine);
+		}
+		$aPlace['address_details'] = $aAddressDetails;
+	}
```
Comment (lonvia) [inline lib/template/details-json.php]: missing: else { $aAddressLines = false; }

Without that you get an undefined variable error later. Dito for $aLinkedPlaces, $aParentOfLines below.

**Comment by lonvia** [comment]
Some general comments:

  * Please have a look at the apache error logs when testing your code. There are many PHP errors thrown which you should resolve.
  * distance, admin_level and is_address need to be included in the address_details output
  * distance and admin_level are also missing for child and parent places
  * polygon output would be needed as well for an HTML frontend

File: lib/template/details-html.php (hunk 1 above)
Comment (agunnarsson) [inline lib/template/details-html.php]: Ok! Then originally tagged values should have a different title, like 'Wikipedia'?

**Comment by agunnarsson** [comment]
Will have a look on the commented issues and get back with an update as soon as I can find some time.

**Comment by lonvia** [comment]
Closing in favour of #968.

**Pull request closed without merging**